File: src/layers/OutputMLP.py
import numpy as np
import math
import tensorflow as tf

class OutputMLP(object):
    def __init__(self):# num_osc, num_h, num_out, init = 'random'):
        self.num_osc = 20
        self.num_h = 20
        self.num_out = 2
        self.init = 'random'
        self.sig = np.vectorize(self._sigmoid)
        self.relu = np.vectorize(self._relu)
        self.relugrad = np.vectorize(self._relugrad)
        
    def _random_mat(self, s1, s2):
        return np.random.random((s1, s2))
    
    def build(self, num_osc, num_h, num_out, a, init = 'random'):
        self.a = a
        self.num_osc = num_osc
        self.num_h = num_h
        self.num_out = num_out
        self.init = init
        if self.init == 'random':
            self.W1 = self._random_mat(self.num_h, self.num_osc) + 1j*self._random_mat(self.num_h, self.num_osc)
            self.W2 = self._random_mat(self.num_out, self.num_h) + 1j*self._random_mat(self.num_out, self.num_h)
        elif self.init == 'zeros':
            self.W1 = np.zeros(
                shape = (self.num_h, self.num_osc),
                dtype = np.complex128) 
            self.w2 = np.zeros(
                shape = (self.num_out, self.num_h),
                dtype = np.complex128)
        else:
            raise ValueError('initialization strategey can only be either random or zeros for {init}'.format(init = self.init))

    def _sigmoid(self, X):
        if math.isclose(X, 0.0, rel_tol = 1e-5, abs_tol = 1e-5):
            return 0.0
        return 1/(1+np.exp(-X))
    
    def sigmoidf(self, x):
        #return self.sig(x)
        return -1+2/(1+np.exp(-self.a*x))
    
    def _relu(self, x):
        if x<0:
            return 0
        return x

    def reluf(self, X):
        return self.relu(X) 

    def set_W1(self, w1):
        self.W1 = w1

    def set_W2(self, w2):
        self.W2 = w2

    def backprop_sigmoid(self, yr, Z, data, lr):
        dW2r = np.matmul(
            -1*np.multiply(
                (data.signal[:, 1:self.num_out+1].T-yr),
                np.multiply(
                    (1+yr),
                    (1-yr)
                )*self.a/2
            ), 
            self.xhr.T
        )
        dW2i = np.matmul(
            np.multiply(
                (data.signal[:, 1:self.num_out+1].T-yr),
                np.multiply(
                    (1+yr),
                    (1-yr)
                )*self.a/2
            ), 
            self.xhi.T
        )
        dW1r = np.matmul(
            -1*np.multiply(
                np.matmul(
                    self.W2.real.T, 
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        np.multiply(
                            (1+yr),
                            (1-yr)
                        )*self.a/2
                    )
                ),
                np.multiply(
                    (1+self.xhr),
                    (1-self.xhr)
                )*self.a/2
            ), 
            Z.real.T
        ) + np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.imag.T, 
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        np.multiply(
                            (1+yr),
                            (1-yr)
                        )*self.a/2 
                    )   
                ),  
                np.multiply(
                    (1+self.xhi),
                    (1-self.xhi)
                )*self.a/2 
            ),  
            Z.imag.T
        )
        dW1i = -1*np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.real.T,
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        np.multiply(
                            (1+yr),
                            (1-yr)
                        )*self.a/2
                    )
                ),
                np.multiply(
                    (1+self.xhr),
                    (1-self.xhr)
                )*self.a/2
            ),
            Z.imag.T
        ) + np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.imag.T,
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        np.multiply(
                            (1+yr),
                            (1-yr)
                        )*self.a/2
                    )
                ),
                np.multiply(
                    (1+self.xhi),
                    (1-self.xhi)
                )*self.a/2
            ),
            Z.real.T
        ) 
        W2r = self.W2.real - lr*dW2r
        #print('dW2r')
        #print(np.max(dW2r))
        #print(np.min(dW2r))
        W2i = self.W2.imag - lr*dW2i
        #print('dW2i')
        #print(np.max(dW2i))
        #print(np.min(dW2i))
        self.set_W2(W2r + 1j*W2i)
        W1r = self.W1.real - lr*dW1r
        #print('dW1r')
        #print(np.max(dW1r))
        #print(np.min(dW1r))
        W1i = self.W1.imag - lr*dW1i
        #print('dW1i')
        #print(np.max(dW1i))
        #print(np.min(dW1i))
        self.set_W1(W1r+1j*W1i)

    def _relugrad(self, x):
        if x>=0:
            return 1
        else:
            return 0
    
    def relugradf(self, x):
        return self.relugrad(x)
    
    def backprop_relu(self, yr, Z, data, lr):
        dW2r = -1*np.matmul(
            np.multiply(
                (data.signal[:, 1:self.num_out+1].T-yr),
                self.relugradf(yr)
            ),  
            self.xhr.T
        )   
        dW2i = np.matmul(
            np.multiply(
                (data.signal[:, 1:self.num_out+1].T-yr),
                self.relugradf(yr)
            ),  
            self.xhi.T
        )
        dW1r = -1*np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.real.T,
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        self.relugradf(yr)
                    )
                ),
                self.relugradf(self.xhr)
            ),
            Z.T.real
        ) + np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.imag.T,
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        self.relugradf(yr)
                    )
                ),
                self.relugradf(self.xhi)
            ),
            Z.T.imag
        )
        dW1i = -1*np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.T.real,
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        self.relugradf(yr)
                    )
                ),
                self.relugradf(self.xhi)
            ),
            Z.imag.T
        ) + np.matmul(
            np.multiply(
                np.matmul(
                    self.W2.T.imag,
                    np.multiply(
                        (data.signal[:, 1:self.num_out+1].T-yr),
                        self.relugradf(yr)
                    )
                ),
                self.relugradf(self.xhr)
            ),
            Z.real.T
        )
        W2r = self.W2.real - lr*dW2r
        print('dW2r')
        print(np.max(dW2r))
        print(np.min(dW2r))
        W2i = self.W2.imag - lr*dW2i
        print('dW2i')
        print(np.max(dW2i))
        print(np.min(dW2i))
        self.set_W2(W2r + 1j*W2i)
        W1r = self.W1.real - lr*dW1r
        print('dW1r')
        print(np.max(dW1r))
        print(np.min(dW1r))
        W1i = self.W1.imag - lr*dW1i
        print('dW1i')
        print(np.max(dW1i))
        print(np.min(dW1i))
        self.set_W1(W1r+1j*W1i)
    

    def __call__(self, Z, activation):
        """
            Z np.ndarray (num_osc, N) The fourier component signals of the output
        """  
        #print(Z) 
        #print('weights')
        #print(self.W1)
        self.nh = np.matmul(self.W1, Z)
        self.nhr = self.nh.real
        self.nhi = self.nh.imag
        self.xhr = activation(self.nhr)
        self.xhi = activation(self.nhi)
        self.xh = self.xhr+1j*self.xhi
        #print(self.xh)
        self.no = np.matmul(self.W2, self.xh)
        self.nor = self.no.real
        self.noi = self.no.imag
        #"""
        self.yr = activation(self.nor)
        self.yi = activation(self.noi)
        """
        self.yr = self.reluf(self.nor)
        self.yi = self.reluf(self.noi)
        #"""
        self.y = self.yr + 1j*self.yi
        return self.yr
       

File: src/cpp_model/source/OutputMLP.cpp
#include "OutputMLP.h"

OutputMLP::OutputMLP(int n_o, int n_h, int n_out, int n, float LR){
    num_osc = n_o;
    num_h = n_h;
    num_out = n_out;
    lr = LR;
    N = n;
    W1 = new std::complex<float> *[num_h];
    Z_h = new std::complex<float> *[num_h];
    Y_h = new std::complex<float> *[num_h];
    for(int i=0; i<num_h; i++){
        W1[i] = new std::complex<float> [num_osc];
        Z_h[i] = new std::complex<float> [N];
        Y_h[i] = new std::complex<float> [N];
        for(int j=0; j<num_osc; j++)
            W1[i][j] = get_complex_random();
    }
    Z_out = new std::complex<float> *[num_out];
    Y = new std::complex<float> *[num_out];
    W2 = new std::complex<float> *[num_out];
    for(int i=0; i<num_out; i++){
        W2[i] = new std::complex<float>[num_h];
        Z_out[i] = new std::complex<float>[N];
        Y[i] = new std::complex<float>[N];
        for(int j=0; j<num_h; j++){
            W2[i][j] = get_complex_random();
        }
    }
}

void OutputMLP::inputLayerOutput(std::complex<float> **X){
    ParamsSigmoid params = {N, 1.0, 0.5};
    std::complex<float> z(0,1);
    for(int i=0; i<num_h; i++){
        for(int j=0; j<num_osc; j++){
            for(int n=0; n<N; n++){
                Z_h[i][n] += W1[i][j]*X[j][n];
            }
        }
        activation.sigmoidf(Z_h[i], Y_h[i], params);        
    }
}

void OutputMLP::hiddenLayerOutput(std::complex<float> **X){
    ParamsSigmoid params = {N, 1.0, 0.5};
    for(int i=0; i<num_out; i++){
        for(int n = 0; n<N; n++){
            for(int j = 0; j<num_h; j++){
                Z_out[i][n] += W2[i][j]*X[j][n];
            }
        }
        activation.sigmoidf(Z_out[i], Y[i], params);
    }
}

std::complex<float>** OutputMLP::forwardPropagation(std::complex<float> **X){
    inputLayerOutput(X);
    hiddenLayerOutput(Y_h);
    return Y; 
}

void OutputMLP::backwardPropagation(float **signal, std::complex<float> **X){
    std::complex<float> iota(0,1);
    float temp1;
    float a = 0.5;
    for(int i=0; i<num_out; i++){
        for(int j=0; j<num_h; j++){ 
            for(int k=0; k<N; k++){
                temp1 = (
                    signal[i][k]-Y[i][k].real()
                )*a*(
                    1-Y[i][k].real()
                )*(
                    1+Y[i][k].real()
                );
                W2[i][j] -= -lr*temp1*(
                    Y_h[j][k].real()-iota*Y_h[j][k].imag()
                );
                for(int l=0; l<num_osc; l++){
                    W1[j][l] -= -lr*temp1*(
                        (
                            W2[i][j].real()*a*(
                                1-Y_h[j][k].real()
                            )*(
                                1+Y_h[j][k].real()
                            )*(
                                X[l][k].real() + iota*X[l][k].imag()
                            ) - W2[i][j].imag()*a*(
                                1-Y_h[j][k].imag()
                            )*(
                                1+Y_h[j][k].imag()
                            )*(
                                X[l][k].imag() + iota*X[l][k].real()
                            )
                        )
                    );
                }
            }
        }
    }
}
